### PR TITLE
404 add methods to pmat for storing prediction into ipfs and adding that to contract mapping

### PR DESCRIPTION
**Pull request by gabrielfior**

### CODE REVIEW

File: prediction_market_agent_tooling/tools/ipfs/ipfs_handler.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+from prediction_market_agent_tooling.config import APIKeys
+from prediction_market_agent_tooling.gtypes import IPFSCIDVersion0
+
+
+class IPFSHandler:
+    def __init__(self, api_keys: APIKeys):
+        self.pinata = PinataPy(
+            api_keys.pinata_api_key.get_secret_value(),
+            api_keys.pinata_api_secret.get_secret_value(),
+        )
```
Comment (gabrielfior) [inline prediction_market_agent_tooling/tools/ipfs/ipfs_handler.py:12]: See https://github.com/gnosis/prediction-market-agent/issues/489

File: tests/markets/omen/test_omen_subgraph_handler.py [hunk 2]
```diff
@@ -285,44 +285,47 @@ def build_incomplete_user_position_from_condition_ids(
     )
 
 
-def test_get_markets_id_in() -> None:
+def test_get_markets_id_in(omen_subgraph_handler: OmenSubgraphHandler) -> None:
     market_id = "0x934b9f379dd9d8850e468df707d58711da2966cd"
-    sgh = OmenSubgraphHandler()
-    markets = sgh.get_omen_binary_markets(
+    markets = omen_subgraph_handler.get_omen_binary_markets(
```
Comment (gabrielfior) [inline tests/markets/omen/test_omen_subgraph_handler.py:290]: Using pytest.fixture here instead of instantiating a new instance every time. Small tweak hence doing it here also.

**Comment by gabrielfior** [comment]
Closes https://github.com/gnosis/prediction-market-agent-tooling/issues/363

File: tests/markets/omen/test_omen_subgraph_handler.py (hunk 2 above)
Comment (gabrielfior) [reply tests/markets/omen/test_omen_subgraph_handler.py:290]: Regardless if it's a singleton or not, we want to instantiate (or get a reference to the created instance) in just one place, and that place is the `conftest`.

File: prediction_market_agent_tooling/deploy/agent.py [hunk 3]
```diff
@@ -476,18 +488,57 @@ def process_market(
...
+    def store_prediction(
+        self, market_id: str, processed_market: ProcessedMarket, keys: APIKeys
+    ) -> None:
+        reasoning = processed_market.answer.reasoning if processed_market.answer.reasoning else ""
+        ipfs_hash = self.ipfs_handler.store_agent_result(
+            IPFSAgentResult(reasoning=reasoning)
+        )
+
+        # We store only the first trade, since it's only here for tracking purposes - also,
+        # we want only 1 entry per prediction, not 1 entry per trade.
```
[the remark this answers is not in the document]
Comment (gabrielfior) [reply prediction_market_agent_tooling/deploy/agent.py]: I think this should be fixed now (https://github.com/gnosis/prediction-market-agent-tooling/pull/446/files#diff-6aaf85d952cb99bbd129bef760f5e94cd4477d698103727006d150de8ff9780fR543) that we store multiple `txHashes` per prediction.

File: tests/markets/omen/test_omen_subgraph_handler.py (hunk 2 above)
Comment (kongzii) [reply tests/markets/omen/test_omen_subgraph_handler.py:290]: > we want to instantiate (or get a reference to the created instance) in just one place

Why? Wasn't your point back then about making it a singleton that it's initialized only once? And no matter how many times we do `OmenSubgraphHandler()` it's always returning the already initialised one? So, is not initialising it "again" just a micro-optimization at best? (because it just returns existing, already initialised, instance)

(Just to be clear: I don't mind refactoring it to a fixture that's passed as an argument as you are doing it here, just trying to understand the motivation as it feels like we are discussing "do not initialize OmenSubgraphHandler" multiple times recently and I thought it was solved nicely by you when it became a singleton. If nothing else, it's nice not to do `OmenSubgraphHandler()` all over the tests just for esthetic reasons 😄 )

File: prediction_market_agent_tooling/deploy/agent.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gabrielfior) [reply prediction_market_agent_tooling/deploy/agent.py]: test

File: tests/markets/omen/test_omen_subgraph_handler.py (hunk 2 above)
Comment (gabrielfior) [reply tests/markets/omen/test_omen_subgraph_handler.py:290]: The point here is not `is it a singleton?` (although for the record it is) - the point is, `fixtures` (https://docs.pytest.org/en/stable/explanation/fixtures.html#about-fixtures) is the pytest way to abstract away the `arrange` phase of the tests. Instantiating the `OmenSubgraphHandler` is necessary for every test and should be done in the same way (e.g. with the same args) on every test, hence it makes sense to abstract it away, and in pytest one does that with fixtures.

File: prediction_market_agent_tooling/markets/omen/omen_subgraph_handler.py [hunk 4]
```diff
@@ -96,6 +99,13 @@ def __init__(self) -> None:
             )
         )
 
+        # ToDo - Update URL once curations are live
```
Comment (evangriffiths) [inline prediction_market_agent_tooling/markets/omen/omen_subgraph_handler.py:102]: Can this TODO be removed now?

File: tests/markets/test_betting_strategies.py [hunk 5]
```diff
@@ -281,7 +281,7 @@ def test_zero_bets() -> None:
         target_p_yes=market.current_p_yes,
         fee=market.fee,
     )
-    assert np.isclose(market_moving_bet.size, 0.0, atol=1e-4)
+    assert np.isclose(market_moving_bet.size, 0.0, atol=1e-3)
```
Comment (evangriffiths) [inline tests/markets/test_betting_strategies.py:284]: eek! flaky test?

File: tests_integration/tools/ipfs/test_ipfs_handler.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+    # We add the current datetime to avoid uploading an existing file (CID is content-based)
+    temp_string = f"Hello World {datetime.datetime.utcnow()}"
+    with NamedTemporaryFile() as temp_file:
+        temp_file.write(temp_string.encode("utf-8"))
+        temp_file.flush()
+        ipfs_hash = test_ipfs_handler.upload_file(temp_file.name)
+
+    # assert uploaded
+    # can take a while to be available for download
+    # r = requests.get(f"https://ipfs.io/ipfs/{ipfs_hash}", timeout=120)
```
Comment (evangriffiths) [inline tests_integration/tools/ipfs/test_ipfs_handler.py:28]:
```suggestion
```

File: tests_integration_with_local_chain/conftest.py [hunk 7]
```diff
@@ -73,3 +76,16 @@ def fund_account_on_tenderly(
     }
     response = requests.post(fork_rpc, json=payload)
     response.raise_for_status()
+
+
+def create_and_fund_random_account(
```
Comment (evangriffiths) [inline tests_integration_with_local_chain/conftest.py:81]: This looks unrelated to the rest of the PR. I see you've added a test for it, but it's not used anywhere in the code. No biggie but would make sense in a separate PR

**Review by evangriffiths — APPROVED** [review]

File: tests_integration_with_local_chain/conftest.py (hunk 7 above)
Comment (gabrielfior) [reply tests_integration_with_local_chain/conftest.py:81]: That's correct, but this test was used to test a part of the function that fails sometimes during CI runs - so I added it here as well for improving test coverage.
Happy to refactor if this becomes too big to maintain or unreliable.

File: tests/markets/test_betting_strategies.py (hunk 5 above)
Comment (gabrielfior) [reply tests/markets/test_betting_strategies.py:284]: I had to increase the absolute tolerance - for some markets, it fails, for others, it works. Should be still a valuable test though.
